pkg/softtainter/softtainter.go

```diff
@@ -0,0 +1,532 @@
+package softtainter
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	promapi "github.com/prometheus/client_golang/api"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/rest"
+	"k8s.io/kubernetes/pkg/util/taints"
+
+	desv1 "github.com/openshift/cluster-kube-descheduler-operator/pkg/apis/descheduler/v1"
+	"github.com/openshift/cluster-kube-descheduler-operator/pkg/operator/operatorclient"
+	"sigs.k8s.io/descheduler/pkg/api"
+	"sigs.k8s.io/descheduler/pkg/descheduler"
+	desclient "sigs.k8s.io/descheduler/pkg/descheduler/client"
+	nodeutil "sigs.k8s.io/descheduler/pkg/descheduler/node"
+	"sigs.k8s.io/descheduler/pkg/framework/pluginregistry"
+	"sigs.k8s.io/descheduler/pkg/framework/plugins/nodeutilization"
+	"sigs.k8s.io/descheduler/pkg/framework/plugins/nodeutilization/classifier"
+	"sigs.k8s.io/descheduler/pkg/framework/plugins/nodeutilization/normalizer"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	prometheusAuthTokenSecretKey        = "prometheusAuthToken"
+	OverUtilizedSoftTaintKey            = "nodeutilization.descheduler.kubernetes.io/overutilized"
+	OverUtilizedSoftTaintValue          = "true"
+	AppropriatelyUtilizedSoftTaintKey   = "nodeutilization.descheduler.kubernetes.io/appropriate"
+	AppropriatelyUtilizedSoftTaintValue = "true"
+)
+
+var (
+	log = logf.Log.WithName("controller_nodeClassifier")
+)
+
+type softTainterArgs struct {
+	useDeviationThresholds bool
+	thresholds             api.ResourceThresholds
+	targetThresholds       api.ResourceThresholds
+	applySoftTaints        bool
+	mode                   desv1.Mode
+}
+
+type softTainter struct {
+	args                              *softTainterArgs
+	resourceNames                     []corev1.ResourceName
+	previousPrometheusClientTransport *http.Transport
+	currentPrometheusAuthToken        string
+	promClient                        promapi.Client
+	promQuery                         string
+	promURL                           string
+	client                            client.Client
+	resyncPeriod                      time.Duration
+	policyConfigFile                  string
+	nodeUtilizationFactory            func(promapi.Client, string) NodeUtilization
+}
+
+type NodeUtilization interface {
+	NodeUsageFromPrometheusMetrics(ctx context.Context) (map[string]map[corev1.ResourceName]*resource.Quantity, error)
+}
+
+type ActualNodeUtilization struct {
```

Comment:
Is this type used outside of the package? If not let's keep it private to avoid any extra support&maintenance.

Reply:
done